File: DSA/Tree/Code/InorderTraversal.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int data;
    struct Node* left;
    struct Node* right;
} Node;

Node* createNode(int data) {
    Node* node = (Node*)malloc(sizeof(Node));
    node->data = data;
    node->left = NULL;
    node->right = NULL;
    return node;
}

// Recursive inorder traversal: Left-Root-Right
void inorderRecursive(Node* root) {
    if (root == NULL) return;
    
    inorderRecursive(root->left);
    printf("%d ", root->data);
    inorderRecursive(root->right);
}

// Iterative inorder traversal using stack
typedef struct {
    Node* data[100];
    int top;
} Stack;

Stack* createStack() {
    Stack* stack = (Stack*)malloc(sizeof(Stack));
    stack->top = -1;
    return stack;
}

void push(Stack* stack, Node* node) {
    if (stack->top < 99) {
        stack->data[++stack->top] = node;
    }
}

Node* pop(Stack* stack) {
    if (stack->top >= 0) {
        return stack->data[stack->top--];
    }
    return NULL;
}

int isEmpty(Stack* stack) {
    return stack->top == -1;
}

void inorderIterative(Node* root) {
    Stack* stack = createStack();
    Node* current = root;
    
    while (current != NULL || !isEmpty(stack)) {
        while (current != NULL) {
            push(stack, current);
            current = current->left;
        }
        
        current = pop(stack);
        printf("%d ", current->data);
        current = current->right;
    }
}

int main() {
    printf("=== Inorder Traversal (C) ===\n\n");
    
    // Create tree
    Node* root = createNode(1);
    root->left = createNode(2);
    root->right = createNode(3);
    root->left->left = createNode(4);
    root->left->right = createNode(5);
    root->right->left = createNode(6);
    root->right->right = createNode(7);
    
    printf("Tree structure:\n");
    printf("         1\n");
    printf("        / \\\n");
    printf("       2   3\n");
    printf("      / \\ / \\\n");
    printf("     4  5 6  7\n\n");
    
    printf("Recursive Inorder (Left-Root-Right): ");
    inorderRecursive(root);
    printf("\n");
    
    printf("Iterative Inorder (using stack):     ");
    inorderIterative(root);
    printf("\n\n");
    
    printf("=== Inorder Characteristics ===\n");
    printf("1. Left subtree → Root → Right subtree\n");
    printf("2. For BST: Produces sorted sequence\n");
    printf("3. Recursive approach: Clean and simple\n");
    printf("4. Iterative approach: Explicit stack management\n\n");
    
    printf("=== Complexity Analysis ===\n");
    printf("Time Complexity:  O(n) - visit each node once\n");
    printf("Space Complexity: O(h) - h = height (recursion/stack)\n");
    printf("                   O(n) worst case (skewed tree)\n");
    
    return 0;
}

File: DSA/Tree/Code/InorderTraversal.py
# Inorder Traversal in Python

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class InorderTraversal:
    @staticmethod
    def inorder_recursive(root):
        """Recursive inorder: Left-Root-Right"""
        if root is None:
            return
        
        InorderTraversal.inorder_recursive(root.left)
        print(root.data, end=" ")
        InorderTraversal.inorder_recursive(root.right)
    
    @staticmethod
    def inorder_iterative(root):
        """Iterative inorder using stack"""
        stack = []
        current = root
        
        while current or stack:
            while current:
                stack.append(current)
                current = current.left
            
            current = stack.pop()
            print(current.data, end=" ")
            current = current.right

# Test
print("=== Inorder Traversal (Python) ===\n")

# Create tree
root = Node(1)
root.left = Node(2)
root.right = Node(3)
root.left.left = Node(4)
root.left.right = Node(5)
root.right.left = Node(6)
root.right.right = Node(7)

print("Tree structure:")
print("         1")
print("        / \\")
print("       2   3")
print("      / \\ / \\")
print("     4  5 6  7\n")

print("Recursive Inorder (Left-Root-Right): ", end="")
InorderTraversal.inorder_recursive(root)
print()

print("Iterative Inorder (using stack):     ", end="")
InorderTraversal.inorder_iterative(root)
print("\n")

print("=== Inorder Characteristics ===")
print("1. Left subtree → Root → Right subtree")
print("2. For BST: Produces sorted sequence")
print("3. Recursive approach: Clean and simple")
print("4. Iterative approach: Explicit stack management\n")

print("=== Complexity Analysis ===")
print("Time Complexity:  O(n) - visit each node once")
print("Space Complexity: O(h) - h = height (recursion/stack)")
print("                   O(n) worst case (skewed tree)")
